### DISCUSSION TITLE
Errorfractionmarker: how does it work?

### DISCUSSION THREAD

**Discussion by tqcthai** [idaholab/moose#26119] [Q&A General] [answered]
Dear all,

I try to understand how ARM work with the errrorfractionmarker. There is an example here, but I don't get it
https://mooseframework.inl.gov/source/markers/ErrorFractionMarker.html

According to the example, we have total error 89, refine fraction is 0.6. As I understand, the elements have the error larger than 89x0.6=53.4 will be refined which shouldn't be elements #7, #9 and #6 as their corresponding errors are under 53.4. Similarly for the coarsen ratio 0.1, we have 89x0.1=8.9 and elements #1, #2, #4, #8, #10 should be coarsen in this case as their errors smaller than 8.9. Am I correct?

Another question is about level 2 and higher refinement levels for the transient problem. After being level 1 refinement, are the sub-elements are treated as the parents element in the above framework of computing error and sorting?

Thanks,
T

**Comment by lindsayad** [comment]
You add elements to refine until your aggregate error reaches 53.4 Similarly you add elements to coarsen until you reach an aggregate error of 8.9

Active elements (regardless of level) are the only elements considered for computing error and for doing refinement/coarsening. Does that answer your question?

**Reply by tqcthai** [reply]
In the example on the above link, we have both coarsen and refine ratios. According to your answer, what aggregate errors that the system follows? I guess the aggregate error you indicated is the average error of the whole domain which is unique.

```
[Adaptivity]
  [Indicators]
    [error]
      type = AnalyticalIndicator
      variable = u
      function = solution
    []
  []
  [Markers]
    [marker]
      type = ErrorFractionMarker
      coarsen = 0.1
      indicator = error
      refine = 0.3
    []
  []
[]
```

**Reply by GiudGiud** [reply]
The aggregate error of the `error` indicator. Look at the sum of that variable over the whole mesh

**Reply by tqcthai** [reply]
Ok, so it's 89 in the illustrated example. As the explanation from Alex, if I use the refine ratio 0.001, the mesh will generate more refined elements than the refine ratio 0.1 because the expecting error of 0.001 is smaller than the case of 0.1 . Am I correct?

**Reply by GiudGiud** [reply, marked as answer]
no you will get less refinement
```
The elements comprising the "refine" fraction, from highest error to lowest, of the total error are marked for refinement.
```
so you will only refine the elements accounting for 1 thousandth of the error as opposed to 10% of the error

**Reply by keassyguang** [reply]
I am also a bit confused about how it works. I checked the source code of `ErrorFractionMarker` and two important methods are shown below. According to the implementation, firstly the maximum and minimum errors of the whole elements are determined through the loop, then the _refine_cutoff and _coarsen_cutoff are calculated. Here `_coarsen_cutoff = _coarsen * _delta + _min` seems to be common in mathematics, however, `_refine_cutoff = (1.0 - _refine) * _max` is a bit weird. It's more intuitive if we define it as `_refine_cutoff = _max - _refine * _delta`.

```
void
ErrorFractionMarker::markerSetup()
{
  if (_clear_extremes)
  {
    _min = std::numeric_limits<Real>::max();
    _max = 0;
  }

  // First find the max and min error
  for (const auto & val : _error_vector)
  {
    _min = std::min(_min, static_cast<Real>(val));
    _max = std::max(_max, static_cast<Real>(val));
  }

  _delta = _max - _min;
  _refine_cutoff = (1.0 - _refine) * _max;
  _coarsen_cutoff = _coarsen * _delta + _min;
}

Marker::MarkerValue
ErrorFractionMarker::computeElementMarker()
{
  Real error = _error_vector[_current_elem->id()];

  if (error > _refine_cutoff)
    return REFINE;
  else if (error < _coarsen_cutoff)
    return COARSEN;

  return DO_NOTHING;
}
```

**Reply by GiudGiud** [reply]
>  however, _refine_cutoff = (1.0 - _refine) * _max is a bit weird.  It's more intuitive if we define it as _refine_cutoff = _max - _refine * _delta.

I think both options make sense in a way, though I do like yours better. Feel free to add a boolean to switch to your option instead
